src/vfio_device.rs

```diff
@@ -933,6 +939,57 @@ impl VfioDevice {
         }
     }
 
+    /// Add all guest memory regions into the vfio device container's iommu table.
+    /// then vfio kernel driver could access guest memory from gfn
+    ///
+    /// # Parameters
+    /// * mem: pinned guest memory which could be accessed by devices binding to the container.
+    pub fn setup_dma_map(&self, mem: &GuestMemoryMmap) -> Result<()> {
+        if !self.iommu_attached {
+            mem.with_regions(|_index, region| {
+                self.container.vfio_dma_map(
+                    region.start_addr().raw_value(),
+                    region.len() as u64,
+                    region.as_ptr() as u64,
+                )
+            })?;
+        }
+        Ok(())
+    }
+
+    /// Remove all guest memory regions from the vfio device container's iommu table.
+    ///
+    /// The vfio kernel driver and device hardware couldn't access this guest memory after
+    /// returning from the function.
+    ///
+    /// # Parameters
+    /// * mem: pinned guest memory which could be accessed by devices binding to the container.
+    pub fn unset_dma_map(&self, mem: &GuestMemoryMmap) -> Result<()> {
+        if !self.iommu_attached {
+            mem.with_regions(|_index, region| {
+                self.container
+                    .vfio_dma_unmap(region.start_addr().raw_value(), region.len() as u64)
+            })?;
+        }
+        Ok(())
+    }
+
+    /// Update the vfio device container's iommu table when guest memory is extended.
+    ///
+    /// # Parameters
+    /// * new_region: extended guest memory region which should be updated.
+    pub fn extend_dma_map(&self, new_region: &Arc<GuestRegionMmap>) -> Result<()> {
```

Comment:
Do we need shrink_dma_map() to undo what's done by extend_dma_map()?

Reply:
Current Cloud Hypervisor only support memory hotplug, for unplug operation, it only takes effect after vm restarted.  I not sure whether we need add the shrink API here?

Reply:
Implementing `shrink_dma_map()` would be nice but it's not needed.
With memory hotplug through ACPI, because the unplug is not supported, there is no need for a shrink API.
With `virtio-mem` support, unplug is supported but `virtio-mem` uses one dedicated region defined upfront for ballooning purposes. Whenever the first hotplug happens, the entire region is hotplugged. But later whenever a new hotplug or unplug happens, it's always part of this region, which means VFIO map does not need to be updated.
I definitely agree we could try to apply some shrinking in order to control at a finer granularity than the region, but this would involve further changes about the overall design. So might be worth keeping this in mind and implementing it later :)